### PR TITLE
Editor: Upgrade TinyMCE to latest version (v4.3.8)

### LINKED ISSUE #4115: Editor: gallery/contact form is duplicated on edit if it appears as first line in editor
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/post/
2. As the very first thing in the editor (no blank lines) add an image gallery (or new contact form)
3. Image gallery shows correctly
4. Click edit, and change image gallery format (eg. circles)

#### What I expected

Image gallery is _modified_ to the new format (eg. circles)

#### What happened instead

Image gallery is _duplicated_ as the new format (eg. circles)

#### Browser / OS version

Chome  49.0.2623.87 OSX 10.11.3 (Could not repro on Firefox latest OSX)

#### Screenshot / Video

edit gallery first line editor

### PR DESCRIPTION
**Pull request by aduth**
Fixes #4115 

This pull request seeks to upgrade our version of TinyMCE from v4.2.8 to its latest version, 4.3.8. Notably, this resolves an issue where attempts to update a gallery or contact form which had been inserted as the first line of the post content were duplicated, rather than replaced. Since the replacement behavior relies on TinyMCE's `editor.execCommand( 'mceInsertContent', false, markup );`, it was suspected and confirmed that an update may resolve the issue. Upon review of [TinyMCE's changelog](https://www.tinymce.com/docs/changelog/) between v4.2.8 and v4.3.8, it's unclear which update brought about the resolution, but as this is a maintenance upgrade we should make anyways, I'm content with its being fixed as a consequence of the upgrade.

__Implementation notes:__

There were two visual regressions as part of the upgrade, resolved within the changes of this pull request:
- The "Toggle Advanced" button was shown as text, rather than as an icon.
 - I opted to replace the icon with its Gridicon equivalent, as we're slowly deprecating Noticon. Because there are already a few other Gridicon-based menu buttons, I refactored their styles to apply more generally to icon buttons.
- The "Paragraph" (format select) text styling had reset to default TinyMCE styling, as a result of a new `.mce-txt` `span` wrapper element being added to its button.

__Upgrade Notes:__

Two changelog notes stuck out as worth exploring:
- [Version 4.3.0](https://www.tinymce.com/docs/changelog/#version430-november232015) added a [`codesample` plugin](https://www.tinymce.com/docs/plugins/codesample/) for syntax highlighting of code blocks, a feature I've personally expressed interest in. However, since it relies on [prism.js](http://prismjs.com/), enabling the plugin would incur additional weight to the page load, so we should at minimum explore options for minimizing this effect (perhaps loading the plugin only when its known the post content contains code?)
- [Version 4.3.4](https://www.tinymce.com/docs/changelog/#version434-february112016) returns a `Promise` instance from `tinymce.init`. Since we already have [post-setup tasks](https://github.com/Automattic/wp-calypso/blob/82ddc3b336901815202101b8e3c506d66883bdfb/client/components/tinymce/index.jsx#L200-L215), we may consider reimplementing this as a promise chain on the `init` call.

__Testing instructions:__

Thoroughly test [post editor](http://calypso.localhost:3000/post) behavior, ensuring no regressions (code or visual). Recommended to test in a non-standard and small-viewport browser. Verify that I've not overlooked changelog revisions which may have unintended consequences.

Related: https://core.trac.wordpress.org/changeset/37000 (cc @azaozz)

/cc @jkudish , @deBhal , @alisterscott

### CODE REVIEW

File: client/components/tinymce/plugins/advanced/plugin.jsx [hunk 1]
```diff
@@ -53,10 +56,16 @@ function advanced( editor ) {
 	editor.addButton( 'wpcom_advanced', {
 		text: 'Toggle Advanced',
 		tooltip: 'Toggle Advanced',
-		classes: 'btn advanced',
+		classes: 'btn wpcom-icon-button advanced',
 		cmd: 'WPCOM_ToggleAdvancedVisible',
 		onPostRender: function() {
 			button = this;
```
Comment (deBhal) [inline client/components/tinymce/plugins/advanced/plugin.jsx]: "button" is getting used in 3 different ways in this function;  It'd be nice to have some help understanding what's going on.  The clearest thing might be to rename this button closure to something like `advancedButton`.  I'm tempted to suggest a comment explaining that we're capturing it here for use in `updateVisibility()`, but I think that will be a lot clearer if it's got a unique name.

Comment (aduth) [inline client/components/tinymce/plugins/advanced/plugin.jsx]: >It'd be nice to have some help understanding what's going on.

I agree, updated naming and added a comment in 90f003b.

**Comment by deBhal** [comment]
I used this branch throughout a support day and didn't see any problems, but I haven't given it the thrashing it deserves.

Code looks good to me, but I'd love to get some more eyes on the editor before we ship.

**Comment by alisterscott** [comment]
I've tested that #4115 has been fixed - works well in Chrome OSX and also checked Firefox and Safari for any regressions. No issues found.

**Comment by aduth** [comment]
Thanks for testing @deBhal and @alisterscott ! In my last pass, I'd discovered (and fixed) one more regression occurring in IE11, one which has come up a few times before (#4402, #4280) and has a relatively straight-forward (but tedious) fix: assigning the width and height directly as styles (included in rebased d52c887).

**Pull request merged**